src/consensus/upgrades.cpp

```diff
@@ -0,0 +1,78 @@
+// Copyright (c) 2018 The Zcash developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include "consensus/upgrades.h"
+
+const struct NUInfo NetworkUpgradeInfo[Consensus::MAX_NETWORK_UPGRADES] = {
+    {
+        /*.nBranchId =*/ 0x54657374,
+        /*.strName =*/ "Test dummy",
+        /*.strInfo =*/ "Test dummy info",
+    }
+};
+
+UpgradeState NetworkUpgradeState(
+    int nHeight,
+    const Consensus::Params& params,
+    Consensus::UpgradeIndex idx)
+{
+    assert(nHeight >= 0);
+    auto nActivationHeight = params.vUpgrades[idx].nActivationHeight;
+
+    if (nActivationHeight == Consensus::NetworkUpgrade::NO_ACTIVATION_HEIGHT) {
+        return UPGRADE_DISABLED;
+    } else if (nHeight >= nActivationHeight) {
+        // From ZIP ???:
+        //
+        // ACTIVATION_HEIGHT
+        //     The block height at which the network upgrade rules will come into effect.
+        //
+        //     For removal of ambiguity, the block at height ACTIVATION_HEIGHT - 1 is
+        //     subject to the pre-upgrade consensus rules.
+        return UPGRADE_ACTIVE;
+    } else {
+        return UPGRADE_PENDING;
+    }
+}
+
+bool NetworkUpgradeActive(
+    int nHeight,
+    const Consensus::Params& params,
+    Consensus::UpgradeIndex idx)
+{
+    return NetworkUpgradeState(nHeight, params, idx) == UPGRADE_ACTIVE;
+}
+
+int CurrentEpoch(int nHeight, const Consensus::Params& params) {
```

Comment:
I didn't bother here because `nHeight` is only used as a parameter to function calls that make the same assumption, so the assertion in the underlying `NetworkUpgradeState()` call will trigger.

Reply:
It's possible that the caller passes in -1 for nHeight.

Reply:
It is possible, but per above comment, `nHeight` is only passed to functions that will eventually have the above `assert`. I can add it here as well if it is a sticking point.